### PR TITLE
add paste.sr.ht

### PR DESCRIPTION
**Pull request by dchinmay2**

### CODE REVIEW

File: fnl/paperplanes/providers/srht.fnl [hunk 1]
```diff
@@ -0,0 +1,34 @@
+(local {: set-field : opts-to-fields} (require :paperplanes.util.providers))
+(local fmt string.format)
+
+(fn make [content-arg meta opts]
+  (local args (doto []
+                    (table.insert :--header)
+                    (table.insert (fmt
+                                    "Authorization:'token %s'"
```
Comment (rktjmp) [inline fnl/paperplanes/providers/srht.fnl]: @p00f Make this "Authorization: token %s", `uv.spawn`/exec will automatically wrap it.

File: fnl/paperplanes/providers/init.fnl [hunk 2]
```diff
@@ -2,4 +2,5 @@
  :ix.io (require :paperplanes.providers.ixio)
  :0x0.st (require :paperplanes.providers.0x0st)
  :sprunge.us (require :paperplanes.providers.sprungeus)
- :paste.rs (require :paperplanes.providers.pasters)}
+ :paste.rs (require :paperplanes.providers.pasters)
+ :srht (require :paperplanes.providers.srht)}
```
Comment (rktjmp) [inline fnl/paperplanes/providers/init.fnl]: Other mirror the url, so probably "sr.ht" or "pastes.sr.ht" (probably a bit much).

File: fnl/paperplanes/providers/srht.fnl (hunk 1 above)
Comment (rktjmp) [reply fnl/paperplanes/providers/srht.fnl]: Also apply similar to the content type header.

Comment (dchinmay2) [reply fnl/paperplanes/providers/srht.fnl]: right, i needed to do that to the json payload too, thanks

File: fnl/paperplanes/providers/init.fnl (hunk 2 above)
Comment (dchinmay2) [reply fnl/paperplanes/providers/init.fnl]: done

**Comment by dchinmay2** [comment]
I only implemented post-string because post-file wasn't actually being used

**Comment by dchinmay2** [comment]
btw I'm not doing anything complicated in the `make` function so how about this:
```fennel
(fn make [content-arg meta opts]
  (local args [:--header
               (fmt "Authorization:token %s" opts.token)
               :--header
               "Content-Type:application/json"
               "https://paste.sr.ht/api/pastes"
               :--data-binary
               (fmt "%s"
                    (vim.json.encode
                      {:visibility (or opts.visibility :unlisted)
                       :files [{:filename (vim.fn.expand "%:t")
                                :contents content-arg}]}))])
```

**Comment by rktjmp** [comment]
Yep, fine. Provider implementation is left to the provider, just has to return the args and response-handler, I just happened to share the `set-field` stuff between the initial few providers so they all build the table iteratively.

**Comment by rktjmp** [comment]
If it's not too much pain, maybe strip/revert the lua files to make a smaller changeset. I don't really have a policy around that with paperplanes but probably makes sense to do it that way.

**Comment by dchinmay2** [comment]
> If it's not too much pain, maybe strip/revert the lua files to make a smaller changeset. I don't really have a policy around that with paperplanes but probably makes sense to do it that way.

I don't understand 😅

File: fnl/paperplanes/providers/srht.fnl [hunk 3]
```diff
@@ -0,0 +1,31 @@
+(local {: set-field : opts-to-fields} (require :paperplanes.util.providers))
```
Comment (rktjmp) [inline fnl/paperplanes/providers/srht.fnl]: Can probably remove these?

File: fnl/paperplanes/providers/srht.fnl [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+(local fmt string.format)
+
+(fn make [content-arg meta opts]
```
Comment (rktjmp) [inline fnl/paperplanes/providers/srht.fnl:3]: Probably add an assert on opts.token so provide a more accurate "you must set provider_opts.token to your sr.ht token" or something?

**Review by rktjmp — COMMENTED** [review]
It's ok, don't worry about the lua. Just two small things in review. Ping back when you're ready to merge.

File: fnl/paperplanes/providers/srht.fnl (hunk 4 above)
Comment (dchinmay2) [reply fnl/paperplanes/providers/srht.fnl:3]: The error message is clear enough in that case
```
Error executing vim.schedule lua callback: vim/_editor.lua:0: paperplanes got no url back from provider: {"errors": [{"reason": "Invalid or expired OAuth token"}]}                                                 
stack traceback:                                                                                                                                                                                                    
        [C]: in function ''                                                                                                                                                                                         
        vim/_editor.lua: in function <vim/_editor.lua:0>  
```

File: fnl/paperplanes/providers/srht.fnl (hunk 3 above)
Comment (dchinmay2) [reply fnl/paperplanes/providers/srht.fnl]: done

File: fnl/paperplanes/providers/srht.fnl (hunk 4 above)
Comment (dchinmay2) [reply fnl/paperplanes/providers/srht.fnl:3]: done

**Pull request merged**